### ISSUE TITLE
Change to use Tuple type in type params

### ISSUE THREAD

**Issue by oxinabox** [invenia/NamedDims.jl#66]
It is a bit uglier but then can dispatch on them.

And the wildcard becomes `Any` 
(I.e  `<:Any` since tuples are covarient)

More generally relax things so it can be any thing allowed in a type-param.
But Symbol is going to be best since then can have kwargs `getindex` etc.

Needed for #43

**Comment by nickrobinson251** [comment]
Any more thoughts on this? 

I'm thinking about if we can tag 1.0, and tempted to suggest yes, unless there are plans to make the change anytime soon.

**Comment by Tokazama** [comment]
Are there any downsides to this besides it being visually less appealing

**Comment by oxinabox** [comment]
No, but it is
significantly less visually appealing.

**Comment by mcabbott** [comment]
Could you clarify what's proposed? It would be possible leave `dimnames(A)` and constructors alone, even if the storage is different, and `:_` is translated to/from `Any`. Then `summary(A)` could print the constructor, like `view` does, for visual appeal.

**Comment by Tokazama** [comment]
I think we can overcome ugly with overloading show but I'm concerned about hurting performance or easy to use syntax for constructing things. Otherwise we get to dispatch on names

**Comment by oxinabox** [comment]
Dispatch on *partial* names.
We can already dispatch on names.
And you almost certainly have to use the ugly form to write it.


I hope we have enough `@ballocated` tests in the package to ensure this does not impact performance

**Comment by mcabbott** [comment]
I guess you could write definitions like`f(x::NamedType(:a, :_)) = "ok"` with a function like `NamedType(sy::Symbol...) = Nda{<:Tuple{map(s -> s===:_ ? Any : s, sy)...}}`.

**Comment by oxinabox** [comment]
An interesting possibility.

**Comment by Tokazama** [comment]
I agree `Tuple{}` isn't as nice as `()` but so far I see this change as a good thing. Once it's set up we should only need to deal with the ugly syntax problem when trying to take advantage of individual names with a `Tuple`. 

That being said, it might be good to let this incubate a bit longer. It's obvious not everyone loves the change and no one has expressed a strong opinion in favor of it. It's always possible that an even better solution is identified or some lurking issue with using `Tuple` is discovered. No need to rewrite this code twice.

**Comment by Tokazama** [comment]
What if we use `Unitful.Dimension`s inside a `Tuple`? I know it would be a bit of a paradigm shift, but it would give us a lot of the capabilities we are looking for here and it's already integrated in so many parts of Julia that it would be easy to get others to adopt extensions to it.
